pkg/model/log.go

```diff
@@ -0,0 +1,12 @@
+package model
+
+import (
+	"context"
+
+	"github.com/puppetlabs/relay-pls/pkg/plspb"
+)
+
+type LogManager interface {
```

Comment:
Along the same lines as the HTTP endpoints, I think it might be a good idea to put in a bit more abstraction here, and also to distance the managers from the HTTP endpoint in naming to make it clear that they are standalone. I'm thinking something like:

```go
type Log struct {
  ID string
  Name string
}

type LogMessage struct {
  Log *Log
  Payload []byte
}

type LogManager interface {
  Create(ctx context.Context, name string) (*Log, error)
  AddMessage(ctx context.Context, logID string, payload []byte) (*LogMessage, error)
}
```

This would make this manager feel consistent with the other managers in this package.

Reply:
This was the way it was originally. I have mixed feelings about this, and it's going to be a repeating concern.

I ended up referencing the PLS objects directly in most of the code. Partly because we (essentially) called for it in the spec.

It's also a uniformly bad idea to duplicate models in multiple places rather than having one centralized source of truth.
And I believe we're going to be better off adding more abstraction and more implementations to the PLS, rather than focusing this within Relay core. Same argument applies to anyone wanting to used an entirely different PLS implementation; it feels like a better option to insist on the PLS contract rather than a vaguely compliant, generic option.

It's a valid argument to somewhat separate just this particular Manager interface. But it comes with all of the standard detractions to duplication.

Reply:
I believe we'll actually make this more explicit, not less. We're going to negate some of the explicit reasons for using gRPC if most of the code follows the existing interface design of the REST API and the Managers. Eventually, I'm leaning towards changing this to a "service" interface and removing this from the Managers altogether. The testing would be altered to focus more on verifying the API at that point, without necessarily adding an explicit "memory" option. It would likely move in the direction of implementing some microservices pattern rather than conform to the existing design at all. And this would still leave this to use the exact gRPC models. The generation and use of these from the proto file is part of the power of gRPC. As I said, this was the way it was coded originally, but it feels like duplicating the same content over and over again for no benefit. And also, as I've previously mentioned, I believe we're better off adding more options in the PLS functionality rather than here, if necessary (i.e a simpler and easier method to store logs rather than the BigQuery-dominated approach for the main path). This would require someone writing their own service to adhere to the interface defined by the PLS, but again, I believe that's an improvement over ambiguity and abstraction here.